From 9251396b886e33618ce2242244c5b36c68788654 Mon Sep 17 00:00:00 2001
From: ywj <ywj@rock-chips.com>
Date: Fri, 23 Nov 2018 10:57:19 +0800
Subject: [PATCH] video: rockchip: vpu: play video happen panic

When the video sesion is in use, the same video sesion is released,
causing a crash, and add locking control

Change-Id: I699493d64870cb1b63b1b036c4eee506f7a204bd
Signed-off-by: Wenjie Yang <ywj@rock-chips.com>
---
 .../video/rockchip/vcodec/vcodec_iommu_ops.c  | 163 +++++++++++-------
 1 file changed, 100 insertions(+), 63 deletions(-)

diff --git a/drivers/video/rockchip/vcodec/vcodec_iommu_ops.c b/drivers/video/rockchip/vcodec/vcodec_iommu_ops.c
index ddb4907a7b59..1ca2418fe99c 100644
--- a/drivers/video/rockchip/vcodec/vcodec_iommu_ops.c
+++ b/drivers/video/rockchip/vcodec/vcodec_iommu_ops.c
@@ -35,7 +35,7 @@ struct vcodec_iommu_session_info *vcodec_iommu_get_session_info
 
 int vcodec_iommu_create(struct vcodec_iommu_info *iommu_info)
 {
-	if (!iommu_info || !iommu_info->ops->create)
+	if (!iommu_info || !iommu_info->ops || !iommu_info->ops->create)
 		return -EINVAL;
 
 	return iommu_info->ops->create(iommu_info);
@@ -47,15 +47,17 @@ int vcodec_iommu_alloc(struct vcodec_iommu_info *iommu_info,
 		       unsigned long align)
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
+	int result = -ENOMEM;
 
-	if (!iommu_info || !iommu_info->ops->alloc || !session)
+	if (!iommu_info || !iommu_info->ops ||
+		!iommu_info->ops->alloc || !session)
 		return -EINVAL;
-
+	mutex_lock(&iommu_info->list_mutex);
 	session_info = vcodec_iommu_get_session_info(iommu_info, session);
 	if (!session_info) {
 		session_info = kzalloc(sizeof(*session_info), GFP_KERNEL);
 		if (!session_info)
-			return -ENOMEM;
+			goto ERROR;
 
 		INIT_LIST_HEAD(&session_info->head);
 		INIT_LIST_HEAD(&session_info->buffer_list);
@@ -66,29 +68,31 @@ int vcodec_iommu_alloc(struct vcodec_iommu_info *iommu_info,
 		session_info->dev = iommu_info->dev;
 		session_info->iommu_info = iommu_info;
 		session_info->buffer_nums = 0;
-		mutex_lock(&iommu_info->list_mutex);
 		list_add_tail(&session_info->head, &iommu_info->session_list);
-		mutex_unlock(&iommu_info->list_mutex);
 	}
 
 	session_info->debug_level = iommu_info->debug_level;
-
-	return iommu_info->ops->alloc(session_info, size, align);
+	result = iommu_info->ops->alloc(session_info, size, align);
+ERROR:
+	mutex_unlock(&iommu_info->list_mutex);
+	return result;
 }
 
 int vcodec_iommu_import(struct vcodec_iommu_info *iommu_info,
 			struct vpu_session *session, int fd)
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
+	int result = -ENOMEM;
 
-	if (!iommu_info || !iommu_info->ops->import || !session)
+	if (!iommu_info || !iommu_info->ops ||
+		!iommu_info->ops->import || !session)
 		return -EINVAL;
-
+	mutex_lock(&iommu_info->list_mutex);
 	session_info = vcodec_iommu_get_session_info(iommu_info, session);
 	if (!session_info) {
 		session_info = kzalloc(sizeof(*session_info), GFP_KERNEL);
 		if (!session_info)
-			return -ENOMEM;
+			goto ERROR;
 
 		INIT_LIST_HEAD(&session_info->head);
 		INIT_LIST_HEAD(&session_info->buffer_list);
@@ -99,40 +103,48 @@ int vcodec_iommu_import(struct vcodec_iommu_info *iommu_info,
 		session_info->dev = iommu_info->dev;
 		session_info->iommu_info = iommu_info;
 		session_info->buffer_nums = 0;
-		mutex_lock(&iommu_info->list_mutex);
 		list_add_tail(&session_info->head, &iommu_info->session_list);
-		mutex_unlock(&iommu_info->list_mutex);
 	}
 
 	session_info->debug_level = iommu_info->debug_level;
-
-	return iommu_info->ops->import(session_info, fd);
+	result = iommu_info->ops->import(session_info, fd);
+ERROR:
+	mutex_unlock(&iommu_info->list_mutex);
+	return result;
 }
 
 int vcodec_iommu_free(struct vcodec_iommu_info *iommu_info,
 		      struct vpu_session *session, int idx)
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
+	int result = -EINVAL;
 
-	session_info = vcodec_iommu_get_session_info(iommu_info, session);
-
-	if (!iommu_info || !iommu_info->ops->free || !session_info)
+	if (!iommu_info || !iommu_info->ops || !iommu_info->ops->free)
 		return -EINVAL;
-
-	return iommu_info->ops->free(session_info, idx);
+	mutex_lock(&iommu_info->list_mutex);
+	session_info = vcodec_iommu_get_session_info(iommu_info,
+			session);
+	if (session_info)
+		result = iommu_info->ops->free(session_info, idx);
+	mutex_unlock(&iommu_info->list_mutex);
+	return result;
 }
 
 int vcodec_iommu_free_fd(struct vcodec_iommu_info *iommu_info,
 			 struct vpu_session *session, int fd)
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
+	int result = -EINVAL;
 
-	session_info = vcodec_iommu_get_session_info(iommu_info, session);
-
-	if (!iommu_info || !iommu_info->ops->free_fd || !session_info)
+	if (!iommu_info || !iommu_info->ops || !iommu_info->ops->free_fd)
 		return -EINVAL;
-
-	return iommu_info->ops->free_fd(session_info, fd);
+	mutex_lock(&iommu_info->list_mutex);
+	session_info = vcodec_iommu_get_session_info(iommu_info,
+				session);
+	if (session_info)
+		result = iommu_info->ops->free_fd(session_info, fd);
+	mutex_unlock(&iommu_info->list_mutex);
+	return result;
 }
 
 int vcodec_iommu_map_iommu_with_iova(struct vcodec_iommu_info *iommu_info,
@@ -141,27 +153,38 @@ int vcodec_iommu_map_iommu_with_iova(struct vcodec_iommu_info *iommu_info,
 				     unsigned long size)
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
+	int result = -EINVAL;
 
-	session_info = vcodec_iommu_get_session_info(iommu_info, session);
-
-	if (!iommu_info || !iommu_info->ops->map_iommu || !session_info)
+	if (!iommu_info || !iommu_info->ops ||
+		!iommu_info->ops->map_iommu_with_iova)
 		return -EINVAL;
-
-	return iommu_info->ops->map_iommu_with_iova(session_info,
-						    idx, iova, size);
+	mutex_lock(&iommu_info->list_mutex);
+	session_info = vcodec_iommu_get_session_info(iommu_info,
+				session);
+	if (session_info)
+		result = iommu_info->ops->map_iommu_with_iova
+				(session_info, idx, iova, size);
+	mutex_unlock(&iommu_info->list_mutex);
+	return result;
 }
 
 int vcodec_iommu_unmap_iommu_with_iova(struct vcodec_iommu_info *iommu_info,
 				       void *session, int idx)
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
+	int result = -EINVAL;
 
-	session_info = vcodec_iommu_get_session_info(iommu_info, session);
-
-	if (!iommu_info || !iommu_info->ops->unmap_iommu || !session_info)
+	if (!iommu_info || !iommu_info->ops ||
+		!iommu_info->ops->unmap_iommu_with_iova)
 		return -EINVAL;
-
-	return iommu_info->ops->unmap_iommu_with_iova(session_info, idx);
+	mutex_lock(&iommu_info->list_mutex);
+	session_info = vcodec_iommu_get_session_info(iommu_info,
+				session);
+	if (session_info)
+		result = iommu_info->ops->unmap_iommu_with_iova
+				(session_info, idx);
+	mutex_unlock(&iommu_info->list_mutex);
+	return result;
 }
 
 int vcodec_iommu_map_iommu(struct vcodec_iommu_info *iommu_info,
@@ -170,33 +193,43 @@ int vcodec_iommu_map_iommu(struct vcodec_iommu_info *iommu_info,
 			   unsigned long *size)
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
+	int result = -EINVAL;
 
-	session_info = vcodec_iommu_get_session_info(iommu_info, session);
-
-	if (!iommu_info || !iommu_info->ops->map_iommu || !session_info)
+	if (!iommu_info || !iommu_info->ops ||
+		!iommu_info->ops->map_iommu)
 		return -EINVAL;
-
-	return iommu_info->ops->map_iommu(session_info, idx, iova, size);
+	mutex_lock(&iommu_info->list_mutex);
+	session_info = vcodec_iommu_get_session_info(iommu_info,
+			session);
+	if (session_info)
+		result = iommu_info->ops->map_iommu(session_info,
+				idx, iova, size);
+	mutex_unlock(&iommu_info->list_mutex);
+	return result;
 }
 
 int vcodec_iommu_unmap_iommu(struct vcodec_iommu_info *iommu_info,
 			     struct vpu_session *session, int idx)
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
+	int result = -EINVAL;
 
-	session_info = vcodec_iommu_get_session_info(iommu_info, session);
-
-	if (!iommu_info || !iommu_info->ops->unmap_iommu || !session_info)
+	if (!iommu_info || !iommu_info->ops || !iommu_info->ops->unmap_iommu)
 		return -EINVAL;
-
-	return iommu_info->ops->unmap_iommu(session_info, idx);
+	mutex_lock(&iommu_info->list_mutex);
+	session_info = vcodec_iommu_get_session_info(iommu_info,
+				session);
+	if (session_info)
+		result = iommu_info->ops->unmap_iommu
+				(session_info, idx);
+	mutex_unlock(&iommu_info->list_mutex);
+	return result;
 }
 
 static int vcodec_iommu_destroy(struct vcodec_iommu_info *iommu_info)
 {
-	if (!iommu_info || !iommu_info->ops->destroy)
+	if (!iommu_info || !iommu_info->ops || !iommu_info->ops->destroy)
 		return -EINVAL;
-
 	return iommu_info->ops->destroy(iommu_info);
 }
 
@@ -205,12 +238,15 @@ void vcodec_iommu_dump(struct vcodec_iommu_info *iommu_info,
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
 
-	session_info = vcodec_iommu_get_session_info(iommu_info, session);
-
-	if (!iommu_info || !iommu_info->ops->dump || !session_info)
+	if (!iommu_info || !iommu_info->ops ||
+		!iommu_info->ops->dump)
 		return;
-
-	iommu_info->ops->dump(session_info);
+	mutex_lock(&iommu_info->list_mutex);
+	session_info = vcodec_iommu_get_session_info
+				(iommu_info, session);
+	if (session_info)
+		iommu_info->ops->dump(session_info);
+	mutex_unlock(&iommu_info->list_mutex);
 }
 
 void vcodec_iommu_clear(struct vcodec_iommu_info *iommu_info,
@@ -218,22 +254,23 @@ void vcodec_iommu_clear(struct vcodec_iommu_info *iommu_info,
 {
 	struct vcodec_iommu_session_info *session_info = NULL;
 
-	session_info = vcodec_iommu_get_session_info(iommu_info, session);
-
-	if (!iommu_info || !iommu_info->ops->clear || !session_info)
+	if (!iommu_info || !iommu_info->ops ||
+		!iommu_info->ops->clear)
 		return;
-
-	iommu_info->ops->clear(session_info);
-
 	mutex_lock(&iommu_info->list_mutex);
-	list_del_init(&session_info->head);
-	kfree(session_info);
+	session_info = vcodec_iommu_get_session_info(iommu_info,
+				session);
+	if (session_info) {
+		iommu_info->ops->clear(session_info);
+		list_del_init(&session_info->head);
+		kfree(session_info);
+	}
 	mutex_unlock(&iommu_info->list_mutex);
 }
 
 int vcodec_iommu_attach(struct vcodec_iommu_info *iommu_info)
 {
-	if (!iommu_info || !iommu_info->ops->attach)
+	if (!iommu_info || !iommu_info->ops || !iommu_info->ops->attach)
 		return 0;
 
 	return iommu_info->ops->attach(iommu_info);
@@ -241,7 +278,7 @@ int vcodec_iommu_attach(struct vcodec_iommu_info *iommu_info)
 
 void vcodec_iommu_detach(struct vcodec_iommu_info *iommu_info)
 {
-	if (!iommu_info || !iommu_info->ops->detach)
+	if (!iommu_info || !iommu_info->ops || !iommu_info->ops->detach)
 		return;
 
 	return iommu_info->ops->detach(iommu_info);
-- 
2.35.3

